# RESA: single cell Recurrently Expressed SNV Analysis
Copyright:

## Summary
RESA detects somatic mutation with high precision directly from scRNA-seq data. The key assumption of this method is that cancer cells evolve in a clonal manner and thus expressed somatic mutations have cross-cell recurrence, whereas the noise and artefacts are likely distributed randomly with small probability of recurrence across the cell population. RESA is composed of three main steps: initial variant calling, filtering and labeling, and modeling and refinement.


## RESA workflow
![image](https://user-images.githubusercontent.com/8051136/136512613-99883afe-0ac4-4b87-9602-c965f787f27f.png)

## Installation
### Dependencies
Perl (version >=5)
R >= 3.4
Python >= 3.4

### Install RESA

## Running RESA
1. Variant calling


2. RESA-filter


3. RESA-refine


